### DISCUSSION TITLE
Use cucumber to run integration tests on host system

### DISCUSSION THREAD

**Discussion by Hsilgos** [cucumber/common#2190] [Q&A Everything Else]
Hi, I'm trying to run integration tests on host system with Robolectric and I can't understand how to integrate runners properly.
My best attempt look like: 
```
@RunWith(RobolectricTestRunner.class)
@Config(manifest = Config.NONE)
@CucumberOptions(
        features =
                 {"features"},
        tags = "not @Disable",
        snippets = CucumberOptions.SnippetType.UNDERSCORE,
        monochrome = true)
public class CustomIntegrationTestRunner {
    @Test
    public void runIntegrationTestsOnLocalJVM() {
        Runner r;
        try {
            r = new Cucumber(CustomIntegrationTestRunner.class);
        } catch (InitializationError e) {
            throw new RuntimeException(e);
        }

        JUnitCore c = new JUnitCore();
        Result result = c.run(Request.runner(r));
        if (!result.wasSuccessful()) {
            System.out.println("============ TEST REPORT ============");
            // Some logic to print error
        }
    }
}
```
It's working, but obviously I'm not satisfied how dos it look like in test report, in log and so on. In test report it looks like single test `CustomIntegrationTestRunner.runIntegrationTestsOnLocalJVM` and I want similar report as I see after running on emulator. 
Making `@RunWith(Cucumber.class)` instead of `@RunWith(RobolectricTestRunner.class)` leads to uninitialised Robolectric, contexts and a lot of other stuff.
May be anybody know how to make it workable?

**Comment by mpkorstanje** [comment]
This is a fairly fundamental problem in JUnit 4. Different runners can not be composed.

I would normally suggest looking at implementing Cucumbers `ObjectFactory` or `Backend` interfaces. You could then in theory invoke the same code as the `RobolectricTestRunner`. But I had a brief look at the `RobolectricTestRunner` and think this would be impossible. Robolectric executes a test method in the context of [`sandbox.runOnMainThread`](https://github.com/robolectric/robolectric/blob/e52400ec8f4cd64506f1b14e9c1bfdbbcefebb49/junit/src/main/java/org/robolectric/internal/SandboxTestRunner.java#L253), Cucumber does not have a hook for that.

What you might explore is using the `ParameterizedRobolectricTestRunner` in combination with the [`TestNGCucumberRunner`](https://github.com/cucumber/cucumber-jvm/blob/main/cucumber-testng/src/main/java/io/cucumber/testng/TestNGCucumberRunner.java).  I think it might be possible to do the setup and tear down in the before/after class methods and use `TestNGCucumberRunner.provideScenarios` to provide the scenarios to execute. The final structure should be quite similar to [`AbstractTestNGCucumberTests`](https://github.com/cucumber/cucumber-jvm/blob/main/cucumber-testng/src/main/java/io/cucumber/testng/AbstractTestNGCucumberTests.java#L19).
